### PR TITLE
Spatial modelling example with decoupled sampling

### PR DESCRIPTION
**Pull request by frazane**
This PR adds a new example for spatial modelling using the decoupled sampling introduced in #195. 

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: examples/spatial.pct.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+# %% [markdown]
+# ## Mean function
+# As stated before, we already know that temperature strongly depends on elevation.
+# So why not use it for our mean function? GPJax lets you define custom mean functions
+# easily: simply subclass `AbstractMeanFunction`.
+
+# %%
+from dataclasses import dataclass
+from jaxtyping import Float, Array
+from gpjax.base import param_field
```
Comment (thomaspinder) [inline examples/spatial.pct.py:113]: Would prefer to see these at the top.

File: examples/spatial.pct.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        out = elevation * self.w + self.b
+        return out
+
+
+# %% [markdown]
+# Now we can define our prior. We'll also choose a Gaussian likelihood.
+
+# %%
+mean_function = MeanFunction()
+prior = gpx.Prior(kernel, mean_function)
```
Comment (thomaspinder) [inline examples/spatial.pct.py:132]: Can you make these keyword arguments?

File: examples/spatial.pct.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+likelihood = gpx.Gaussian(D.n)
+
+# %% [markdown]
+# Finally we construct the posterior.
+# %%
+posterior = prior * likelihood
+print(posterior)
```
Comment (thomaspinder) [inline examples/spatial.pct.py:139]: Do we need print here?

File: examples/spatial.pct.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+# %% [markdown]
+# Now we can define our prior. We'll also choose a Gaussian likelihood.
+
+# %%
+mean_function = MeanFunction()
+prior = gpx.Prior(kernel, mean_function)
+likelihood = gpx.Gaussian(D.n)
+
+# %% [markdown]
+# Finally we construct the posterior.
```
Comment (thomaspinder) [inline examples/spatial.pct.py:136]: Finally,

File: examples/spatial.pct.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+# %% [markdown]
+# ## Model fitting
+# We proceed to train our model. Because we used a Gaussian likelihood, the resulting posterior is
+# a `ConjugatePosterior`, which allows us to optimize the analytically expressed marginal loglikelihood.
+#
+# As always, we can jit-compile the objective function to speed things up.
+# %%
+
+negative_mll = jax.jit(gpx.objectives.ConjugateMLL(negative=True))
+negative_mll(posterior, train_data=D)
```
Comment (thomaspinder) [inline examples/spatial.pct.py:151]: Is there a reason you call the objective here?

File: examples/spatial.pct.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+# %%
+import optax as ox
```
Comment (thomaspinder) [inline examples/spatial.pct.py:154]: To the top.

File: examples/spatial.pct.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+posterior, history = gpx.fit(
+    model=posterior,
+    objective=negative_mll,
+    train_data=D,
+    optim=optim,
+    num_iters=3000,
+    safe=True,
+)
+
+posterior: gpx.gps.ConjugatePosterior
```
Comment (thomaspinder) [inline examples/spatial.pct.py:166]: What is this line doing?

File: examples/spatial.pct.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+# %% [markdown]
+# ## Sampling on a grid
+# Now comes the cool part. In a standard GP implementation, for n test points, we have a O(n^2)
```
Comment (thomaspinder) [inline examples/spatial.pct.py:170]: Use LaTeX here i.e., `$\mathcal{O}(n^2)$`

File: examples/spatial.pct.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+# computational complexity and O(n^2) memory requirement. We want to make predictions on a total
+# of roughly 70'000 pixels, and that would require us to compute a covariance matrix of `70000 ** 2 = 4900000000` elements.
+# If these are `float64`s, as it is often the case in GPJax, it would be equivalent to more than 36 Gigabytes of memory. And
```
Comment (thomaspinder) [inline examples/spatial.pct.py:173]: Great motivation :D

File: examples/spatial.pct.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+# that's for a fairly coarse and tiny grid. If we were to make predictions on a 1000x1000 grid, the total memory required
+# would be 8 _Therabytes_ of memory, which is untractable.
```
Comment (thomaspinder) [inline examples/spatial.pct.py:175]: Therabytes -> Terabytes

Comment (thomaspinder) [reply examples/spatial.pct.py:175]: untractable -> intractable.

File: examples/spatial.pct.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+# Fortunately, the pathwise conditioning method allows us to sample from our posterior with linear complexity,
+# O(n), with the number of pixels.
```
Comment (thomaspinder) [inline examples/spatial.pct.py:177]: 1) Latex 2) ~with~ -> in

File: examples/spatial.pct.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+predtest[~mask] = ytest[:, :9]
+predtest = predtest.unstack()
+predtest.plot(
+    col="realization",
+    col_wrap=3,
+    cbar_kwargs={"aspect": 50, "pad": 0.02, "label": "Max. daily temperature [°C]"},
+)
+
+# %% [markdown]
+# Remember when we said that on average the temperature decreases with heigth at a rate
```
Comment (thomaspinder) [inline examples/spatial.pct.py:239]: height

File: examples/spatial.pct.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+# of approximately -6.5°C/km? That's -0.0065°C/m. The `w` parameter of our mean function
+# is very close: we have learned the environmental lapse rate!
```
Comment (thomaspinder) [inline examples/spatial.pct.py:241]: Awesome 🎉

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Awesome work @frazane :D You didn't disappoint!

Please link to your notebook in `docs/index.md` so that it correctly displays in the docs.

File: examples/spatial.pct.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+# Observed temperature data
+temperature = pd.read_csv("data/max_tempeature_switzerland.csv")
+temperature = gpd.GeoDataFrame(
+    temperature,
+    geometry=gpd.points_from_xy(temperature.longitude, temperature.latitude),
+).dropna(how="any")
+
+# Country borders shapefile
+path = "simplecache::https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/10m/cultural/ne_10m_admin_0_countries.zip"
```
[the remark this answers is not in the document]
Comment (frazane) [reply examples/spatial.pct.py:47]: Naturalearthdata has been around for quite some time (well established libraries like cartopy also use their data), should be quite stable. But if you prefer we can download the shapefile and store it in the repo, it's small. Another option is to use cartopy ourselves.

File: examples/spatial.pct.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+import pystac_client
+import rioxarray as rio
+import xarray as xr
+from rioxarray.merge import merge_arrays
+
+import gpjax as gpx
+
+jax.config.update("jax_enable_x64", True)
+
+key = jr.PRNGKey(23908)
```
[the remark this answers is not in the document]
Comment (frazane) [reply examples/spatial.pct.py:37]: ahah just randomly hit some numbers on my keyboard

File: examples/spatial.pct.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (thomaspinder) [reply examples/spatial.pct.py:47]: OK, fine by me. Just wanted to clarify in case of future issues.

File: examples/spatial.pct.py (hunk 7 above)
Comment (frazane) [reply examples/spatial.pct.py:166]: Yeah I was going to bring this up btw. The fit method accepts and returns `Module` types and IDE language servers (e.g. pylance) won't know that `posterior` is of type `ConjugatePosterior` in this case. So no "mouse hovering" for automatic docstring display, no tab completion, no methods suggestions. By adding this line we tell the language server the type. I can remove it for now but would be nice to think of a way (overloading?) to return the same specific type that was passed to `fit`. Just a user-friendliness thing.

Comment (thomaspinder) [reply examples/spatial.pct.py:166]: Neat. I didn't know this was possible. Let's discuss it further on Thursday but, for now, this line makes sense.

**Review by thomaspinder — APPROVED** [review]

**Pull request merged**